### ISSUE TITLE
add optional species, genome, and demographic features to stdpopsim

### ISSUE THREAD

**Issue by fbaumdicker** [popsim-consortium/stdpopsim#848]
Some aspects of species and genomes can be simulated in slim that are not captured in msprime, but it would still be reasonable to simulate the species in both simulators.

A concrete example we have in mind is the circular genome of bacteria.
In msprime we just ignore this such that the evolution at the borders of the genome is not connected to each other.
In slim this can be included.

It might be nice to have a way in stdpopsim to define such genome properties at the species level and maybe also at the demographic model level? In msprime the optional parameter would be ignored , while in slim it would enable the corresponding flag/parameter of the simulation.

There might also be examples in the other direction, where a feature of msprime is not useful/possible in slim.

**Comment by jeromekelleher** [comment]
I guess one thing we could do is attach some attributes to each of the chromosomes, which could cover things like sex chromosomes, circularity, etc. One thing we could do is have a ``flags`` value for each chr, and have some bitwise flags corresponding to the various properties of a chromosome that we might have (say AUTOSOME, SEX_X, CIRCULAR,....). This could be declared in the species.py file.

Or have I been writing too much C, and we should add a bunch of boolean attributes to the Chromosome class?

**Comment by petrelharp** [comment]
> Or have I been writing too much C, and we should add a bunch of boolean attributes to the Chromosome class?

Or maybe an `.attributes` dict that could serve the same purpose as `.flags` but be more readable?

**Comment by jeromekelleher** [comment]
Yeah, if we could restrict the values in that dict to a fixed set of keys.

**Comment by petrelharp** [comment]
Another thing we'd like to add is e.g. "selfing rate", which would be a number.

**Comment by petrelharp** [comment]
As discussed with @jeanrjc, for bacteria we need:

- circular chromosome
- haploid
- HGT (which would be implemented as gene conversion with msprime)

**Comment by petrelharp** [comment]
As mentioned in #861, we should add `ploidy`.

**Comment by petrelharp** [comment]
I think we also really need a `description` field. For instance, see [these notes](https://github.com/popsim-consortium/stdpopsim/pull/826/files#diff-1ea26572750af2149e2b8d1d847bde40a2952d3f133646115ed9d34c2bf9dae4R176).

**Comment by apragsdale** [comment]
> I think we also really need a `description` field. For instance, see [these notes](https://github.com/popsim-consortium/stdpopsim/pull/826/files#diff-1ea26572750af2149e2b8d1d847bde40a2952d3f133646115ed9d34c2bf9dae4R176).

Agreed - having those as a description instead of comments would be helpful (and less liable to be accidentally deleted over time)

**Comment by grahamgower** [comment]
> * type (default="A" or "autosome", other options="X", "Y", "Z", "W", "mitochondria", "chloroplast"... ?)

I think we might need to break this down a bit more, to consider the functionality required for these various types, rather than just having their names. Eg. discussion of sex chromsomes in #383 suggests X needs a parameter `p` (male/female ratio), in order to adjust Ne. Y, Mitochondria and chloroplast might simply need a ploidy parameter.

Similarly, I think the `reproduction_mode` might need to be broken down by the required functionality.

**Comment by jeromekelleher** [comment]
So there's a lot of things being proposed at once here, ranging from the fairly simple and uncontroversial (``Species.ploidy``) to really quite hard and complicated (chromosome type). How about we break this into a separate issues for each, so that we can address them independently?

**Comment by fbaumdicker** [comment]
Yes it makes sense to discuss the more complicated stuff separately. But I think adding a description field might be a simple first solution for many of these issues.
